File: bananapeel.js
// A decoder for the BANANAPEEL algorithm that can run in the browser. This is deliberately minimal, but it will decode any BANANAPEEL message produced
// by a compliant encoder. Since this runs in an interpreted language, do NOT expect this to be fast!

const bpDecode = (key_str, partitions) => {
    // Convert the key string into an actual key
    const binary_key = atob(key_str);
    const key_bytes = new Uint8Array(binary_key.length);
    for (let i = 0; i < binary_key.length; i++) {
        key_bytes[i] = binary_key.charCodeAt(i);
    }
    const key = {
        rng_init_state: new DataView(key_bytes.buffer).getBigUint64(0, true),
        rng_init_seq: new DataView(key_bytes.buffer).getBigUint64(8, true),
        base64_len: new DataView(key_bytes.buffer).getUint32(16, true),
        noise_len: new DataView(key_bytes.buffer).getUint32(20, true),
    };

    // Utility functions for taking moduli in the 64-bit and 32-bit integer ranges
    const wrap64 = x => x % (18446744073709551615n + 1n);
    const wrap32 = x => x % (4294967295n + 1n)

    // Minimal PCG implementation with seeding
    let rng = {
        state: 0n,
        inc: wrap64((key.rng_init_seq << 1n) | 1n),
        next: function() {
            const oldstate = this.state;
            this.state = wrap64((oldstate * 6364136223846793005n) + (this.inc | 1n));
            const xorshifted = wrap32(((oldstate >> 18n) ^ oldstate) >> 27n);
            const rot = wrap32(oldstate >> 59n);
            return Number(wrap32((xorshifted >> rot) | (xorshifted << ((-rot) & 31n))));
        }
    };
    rng.next();
    rng.state = wrap64(rng.state + key.rng_init_state);
    rng.next();

    // Order and parse the partitions
    let next_idx = 0; // The next index to find
    while (next_idx < partitions.length) {
        let order_prefix = rng.next();
        let formatted_order_prefix = order_prefix.toString(16).padStart(8, '0');
        // NOTE: Not worth searching among those we've already parsed obviously
        for (let i = next_idx; i < partitions.length; i++) {
            if (partitions[i].startsWith(formatted_order_prefix)) {
                // Strip off the order prefix and the noise
                let data = partitions[i].substring(formatted_order_prefix.length);
                data = data.substring(key.noise_len);
                partitions[i] = data;
                // Now move that partition to the index we want it at
                [partitions[i], partitions[next_idx]] = [partitions[next_idx], partitions[i]];
                next_idx += 1;
            }
        }
    }

    // Reverse the hex encoding
    let hex = partitions.join("");
    // If there are an odd number of characters, we padded, so get rid of the last character until we handle the padding properly in
    // the base64 (otherwise we'll get an error)
    hex = hex.length % 2 !== 0 ? hex.substring(0, hex.length - 1) : hex;
    let base64 = new Uint8Array(hex.length / 2);
    for (let i = 0; i < hex.length; i += 2) {
        const byte = parseInt(hex.substr(i, 2), 16);
        base64[i / 2] = byte;
    }

    // Remove the base 64 padding added to make the chunks even
    let base64_unpadded = base64.slice(0, key.base64_len);
    // Decode the base 64
    let base64_unpadded_str = new TextDecoder().decode(base64_unpadded);

    // JS needs a base64 *string* in order to perform native decoding (expects URL-safe)
    return atob(base64_unpadded_str);
}

File: bananapeel.py
# A decoder for the BANANAPEEL algorithm that can run in Python. This is deliberately minimal, but it will decode any BANANAPEEL message produced
# by a compliant encoder. Since this runs in an interpreted language, do NOT expect this to be fast!
#
# This code has deliberately been made extremely minimal and compressed.

import sys
import base64
import struct

PCG_MAGIC_MODULUS_1 = 18446744073709551615
PCG_MAGIC_MODULUS_2 = 4294967295

# Read the partitions from stdin
partitions = [line.strip() for line in sys.stdin.readlines()]
# Unpack the key string into its components
key = dict(zip(
    ['rng_init_state', 'rng_init_seq', 'base64_len', 'noise_len'],
    struct.unpack('<QQII', base64.b64decode(sys.argv[1])[:24])
))

# Utility function to handle modulus operation for different integer ranges
wrap = lambda x, limit: x % (limit + 1)

# Minimal PCG implementation with seeding
def rng_next(rng):
    oldstate = rng['state']
    rng['state'] = wrap(oldstate * 6364136223846793005 + (rng['inc'] | 1), PCG_MAGIC_MODULUS_1)
    xorshifted = wrap(((oldstate >> 18) ^ oldstate) >> 27, PCG_MAGIC_MODULUS_2)
    rot = wrap(oldstate >> 59, PCG_MAGIC_MODULUS_2)
    return wrap((xorshifted >> rot) | (xorshifted << (-rot & 31)), PCG_MAGIC_MODULUS_2)

rng = {
    'state': 0,
    'inc': wrap(key['rng_init_seq'] << 1 | 1, PCG_MAGIC_MODULUS_1)
}
rng_next(rng)
rng['state'] = wrap(rng['state'] + key['rng_init_state'], PCG_MAGIC_MODULUS_1)
rng_next(rng)

# Order and parse the partitions based on the RNG
next_idx = 0
while next_idx < len(partitions):
    order_prefix = rng_next(rng)
    formatted_order_prefix = f"{order_prefix:08x}"
    for i in range(next_idx, len(partitions)):
        if partitions[i].startswith(formatted_order_prefix):
            data = partitions[i][len(formatted_order_prefix):][key['noise_len']:]
            partitions[i], partitions[next_idx] = partitions[next_idx], partitions[i]
            partitions[next_idx] = data
            next_idx += 1

# Reconstruct the hex string from partitions and handle padding, then decode the base64 to retrieve the original string
print(base64.urlsafe_b64decode(bytes.fromhex(''.join(partitions)[:-1] if len(''.join(partitions)) % 2 != 0 else ''.join(partitions))[:key['base64_len']]).decode('utf-8'))
